File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/elements/cards/CardSpoiler.java
package com.sup.dev.android.views.elements.cards;

import android.graphics.PorterDuff;
import android.text.Html;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.sup.dev.android.androiddevsup.R;

import java.util.ArrayList;

public class CardSpoiler extends Card {

    private final ArrayList<Card> cards = new ArrayList<>();
    private String title;
    private String text;
    private String rightText;
    private int titleColor = 0x01FF0000;
    private int rightTextColor = 0x01FF0000;
    private int textColor = 0x01FF0000;
    private boolean originalSeted = false;
    private int titleColorOriginal;
    private int rightTextColorOriginal;
    private int textColorOriginal;
    private int iconColor = 0x01FF0000;

    boolean expanded;
    boolean enabled = true;

    @Override
    public int getLayout() {
        return R.layout.card_spoiler;
    }

    @Override
    public void bindView(View view) {
        ImageView vIcon = view.findViewById(R.id.icon);
        TextView vTitle = view.findViewById(R.id.title);
        TextView vText = view.findViewById(R.id.text);
        TextView vRightText = view.findViewById(R.id.right_text);
        View vTouch = view.findViewById(R.id.touch);

        if (!originalSeted) {
            originalSeted = true;
            titleColorOriginal = vTitle.getCurrentTextColor();
            textColorOriginal = vText.getCurrentTextColor();
            rightTextColorOriginal = vRightText.getCurrentTextColor();
        }

        vText.setText(text == null ? null : Html.fromHtml(text));
        vRightText.setText(rightText == null ? null : Html.fromHtml(rightText));
        vTitle.setText(title == null ? null : Html.fromHtml(title));

        vText.setVisibility(text == null ? View.GONE : View.VISIBLE);
        vRightText.setVisibility(rightText == null ? View.GONE : View.VISIBLE);
        vTitle.setVisibility(title == null ? View.GONE : View.VISIBLE);

        vText.setEnabled(enabled);
        vRightText.setEnabled(enabled);
        vTitle.setEnabled(enabled);

        vText.setTextColor(textColor != 0x01FF0000 ? textColor : textColorOriginal);
        vRightText.setTextColor(rightTextColor != 0x01FF0000 ? rightTextColor : rightTextColorOriginal);
        vTitle.setTextColor(titleColor != 0x01FF0000 ? titleColor : titleColorOriginal);

        vIcon.setImageResource(expanded ? R.drawable.ic_keyboard_arrow_up_white_24dp : R.drawable.ic_keyboard_arrow_down_white_24dp);
        vIcon.setAlpha(enabled ? 255 : 106);
        if (iconColor != 0x01FF0000) vIcon.setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP);
        vTouch.setOnClickListener(!enabled ? null : v -> setExpanded(!expanded));
        vTouch.setClickable(enabled);
    }

    //
    //  Setters
    //

    public CardSpoiler add(Card card) {
        cards.add(card);
        return this;
    }

    public CardSpoiler setTitle(String title) {
        this.title = title;
        update();
        return this;
    }

    public CardSpoiler setExpanded(boolean expanded) {
        this.expanded = expanded;
        update();

        if (expanded) {

            int myIndex = adapter.indexOf(this);
            for (Card c : cards)
                if (myIndex != -1)
                    if (!adapter.contains(c)) adapter.add(++myIndex, c);

        } else
            for (Card c : cards) adapter.remove(c);

        return this;
    }

    public CardSpoiler setEnabled(boolean enabled) {
        this.enabled = enabled;
        update();
        return this;
    }

    public CardSpoiler setText(String text) {
        this.text = text;
        update();
        return this;
    }

    public CardSpoiler setRightText(String rightText) {
        this.rightText = rightText;
        update();
        return this;
    }

    public CardSpoiler setIconColor(int iconColor) {
        this.iconColor = iconColor;
        update();
        return this;
    }

    public CardSpoiler setTextColor(int textColor) {
        this.textColor = textColor;
        update();
        return this;
    }

    public CardSpoiler setTitleColor(int titleColor) {
        this.titleColor = titleColor;
        update();
        return this;
    }

    public CardSpoiler setRightTextColor(int rightTextColor) {
        this.rightTextColor = rightTextColor;
        update();
        return this;
    }

    //
    //  Getters
    //


    public ArrayList<Card> getCards() {
        return cards;
    }
}

File: src/main/java/com/sup/dev/android/views/cards/CardSpoiler.kt
package com.sup.dev.android.views.cards

import android.graphics.PorterDuff
import androidx.annotation.StringRes
import android.text.Html
import android.view.Gravity
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.sup.dev.android.R
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.views.support.adapters.CardAdapter
import java.util.ArrayList


open class CardSpoiler : Card(R.layout.card_spoiler) {

    //
    //  Getters
    //

    val cards = ArrayList<Card>()
    private var titleGravity = Gravity.LEFT
    private var title: String? = null
    private var titleExpanded: String? = null
    private var text: String? = null
    private var rightText: String? = null
    private var titleColor = 0
    private var rightTextColor = 0
    private var textColor = 0
    private var originalSeted = false
    private var dividerVisible = true
    private var titleColorOriginal = 0
    private var rightTextColorOriginal = 0
    private var textColorOriginal = 0
    private var iconColor = 0
    private var useExpandedArrow = true
    private var useExpandedTitleArrow = false

    internal var expanded = false
    internal var enabled = true

    @Suppress("DEPRECATION")
    override fun bindView(view: View) {
        super.bindView(view)
        val vIcon:ImageView = view.findViewById(R.id.vIcon)
        val vTitle:TextView = view.findViewById(R.id.vTitle)
        val vText:TextView = view.findViewById(R.id.vText)
        val vRightText:TextView = view.findViewById(R.id.vRightText)
        val vTouch:View = view.findViewById(R.id.vTouch)
        val vDivider:View = view.findViewById(R.id.vDivider)

        val iconDown = ToolsResources.getDrawableAttrId(R.attr.ic_keyboard_arrow_down_24dp)
        val iconUp = ToolsResources.getDrawableAttrId(R.attr.ic_keyboard_arrow_up_24dp)

        if (!originalSeted) {
            originalSeted = true
            titleColorOriginal = vTitle.currentTextColor
            textColorOriginal = vText.currentTextColor
            rightTextColorOriginal = vRightText.currentTextColor
        }

        vText.text = if (text == null) null else Html.fromHtml(text)
        vRightText.text = if (rightText == null) null else Html.fromHtml(rightText)

        if (expanded && titleExpanded != null)
            vTitle.text = Html.fromHtml(titleExpanded)
        else
            vTitle.text = if (title == null) null else Html.fromHtml(title)

        vText.visibility = if (text == null) View.GONE else View.VISIBLE
        vRightText.visibility = if (rightText == null) View.GONE else View.VISIBLE
        vTitle.visibility = if (title == null) View.GONE else View.VISIBLE
        vIcon.visibility = if (useExpandedArrow) View.VISIBLE else View.GONE



       if(useExpandedTitleArrow) vTitle.setCompoundDrawablesWithIntrinsicBounds(0, 0, if (expanded) iconUp else iconDown, 0)
       else vTitle.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)


        vText.isEnabled = enabled
        vRightText.isEnabled = enabled
        vTitle.isEnabled = enabled

        (vTitle.layoutParams as LinearLayout.LayoutParams).gravity = titleGravity

        vDivider.visibility = if (dividerVisible) View.VISIBLE else View.GONE
        vText.setTextColor(if (textColor != 0) textColor else textColorOriginal)
        vRightText.setTextColor(if (rightTextColor != 0) rightTextColor else rightTextColorOriginal)
        vTitle.setTextColor(if (titleColor != 0) titleColor else titleColorOriginal)

        vIcon.setImageResource(if (expanded) iconUp else iconDown)
        vIcon.setAlpha(if (enabled) 255 else 106)
        if (iconColor != 0) vIcon.setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP)
        if (enabled) vTouch.setOnClickListener { setExpanded(!expanded) }
        else vTouch.setOnClickListener(null)
        vTouch.isClickable = enabled
    }

    //
    //  Setters
    //

    fun setUseExpandedArrow(useExpandedArrow:Boolean): CardSpoiler {
        this.useExpandedArrow = useExpandedArrow
        update()
        return this
    }

    fun setUseExpandedTitleArrow(useExpandedArrow:Boolean): CardSpoiler {
        this.useExpandedTitleArrow = useExpandedArrow
        update()
        return this
    }

    fun add(card: Card): CardSpoiler {
        cards.add(card)
        setExpanded(expanded)
        return this
    }

    fun remove(card: Card): CardSpoiler {
        cards.remove(card)
        adapter?.remove(card)
        return this
    }

    open fun setTitle(@StringRes title: Int): CardSpoiler {
        return setTitle(ToolsResources.s(title))
    }

    fun setTitle(title: String?): CardSpoiler {
        this.title = title
        update()
        return this
    }

    open fun setTitleExpanded(@StringRes title: Int): CardSpoiler {
        return setTitleExpanded(ToolsResources.s(title))
    }

    fun setTitleExpanded(title: String?): CardSpoiler {
        this.titleExpanded = title
        update()
        return this
    }

    fun setTitleGravity(gravity: Int): CardSpoiler {
        this.titleGravity = gravity
        update()
        return this
    }

    override fun setCardAdapter(adapter: CardAdapter?) {
        super.adapter = adapter
        setExpanded(expanded)
    }

    open fun onExpandedClicked(expanded: Boolean){

    }

    fun setExpanded(expanded: Boolean): CardSpoiler {
        this.expanded = expanded
        onExpandedClicked(expanded)
        update()

        if (adapter != null) {
            if (expanded) {

                var myIndex = adapter!!.indexOf(this)
                for (c in cards)
                    if (myIndex != -1) {
                        ++myIndex
                        if (!adapter!!.contains(c)) adapter!!.add(myIndex, c)
                    }

            } else
                for (c in cards) adapter!!.remove(c)
        }

        return this
    }

    fun setEnabled(enabled: Boolean): CardSpoiler {
        this.enabled = enabled
        update()
        return this
    }

    fun setText(text: String): CardSpoiler {
        this.text = text
        update()
        return this
    }

    fun setRightText(rightText: String): CardSpoiler {
        this.rightText = rightText
        update()
        return this
    }

    fun setIconColor(iconColor: Int): CardSpoiler {
        this.iconColor = iconColor
        update()
        return this
    }

    fun setTextColor(textColor: Int): CardSpoiler {
        this.textColor = textColor
        update()
        return this
    }

    fun setTitleColor(titleColor: Int): CardSpoiler {
        this.titleColor = titleColor
        update()
        return this
    }

    fun setRightTextColor(rightTextColor: Int): CardSpoiler {
        this.rightTextColor = rightTextColor
        update()
        return this
    }

    fun setDividerVisible(dividerVisible: Boolean): CardSpoiler {
        this.dividerVisible = dividerVisible
        update()
        return this
    }
}
